### PR TITLE
Add ability to authenticate AWS batch with private ECR repositories

### PR DESCRIPTION
**Pull request by jawadqur**
### Improvements
- Add ability to authenticate AWS batch with ECR repositories via user data

### CODE REVIEW

File: hatchery/nextflow.go [hunk 1]
```diff
@@ -526,13 +526,83 @@ func setupVpcAndSquid(ec2Svc *ec2.EC2, userName string, hostname string) (*strin
...
+
+// Create AWS Batch compute environment
 func createBatchComputeEnvironment(userName string, hostname string, tagsMap map[string]*string, batchSvc *batch.Batch, ec2Svc *ec2.EC2, iamSvc *iam.IAM, vpcid string, subnetids []string, payModel *PayModel, awsAccountId string, nextflowConfig NextflowConfig) (string, error) {
 	instanceProfileArn, err := createEcsInstanceProfile(iamSvc, fmt.Sprintf("%s-nf-ecsInstanceRole", hostname))
 	if err != nil {
 		Config.Logger.Printf("Unable to create ECS instance profile: %s", err.Error())
 		return "", err
 	}
 
+	// the launch template for the compute envrionemtn must be user-specific as well
```
Comment (paulineribeyre) [inline hatchery/nextflow.go:600]:
```suggestion
	// the launch template for the compute environment must be user-specific as well
```

File: hatchery/nextflow.go [hunk 2]
```diff
@@ -1474,3 +1558,23 @@ workDir = '%s'`,
...
+// function to generate user data
+func generateUserData(userName string) (string, error) {
+	// TODO: read repo from config
+	approvedRepo := "143731057154.dkr.ecr.us-east-1.amazonaws.com/nextflow-approved"
+
+	// TODO: read region from config
+	userData := base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf(`MIME-Version: 1.0
+Content-Type: multipart/mixed; boundary="==MYBOUNDARY=="
+
+--==MYBOUNDARY==
```
Comment (paulineribeyre) [inline hatchery/nextflow.go:1571]: what does MYBOUNDARY do?

**Review by paulineribeyre — COMMENTED** [review]
thank you this is great!! just some clean up comments

File: hatchery/nextflow.go [hunk 3]
```diff
@@ -551,6 +621,12 @@ func createBatchComputeEnvironment(userName string, hostname string, tagsMap map
 		Config.Logger.Printf("Debug: Batch compute environment '%s' already exists, updating it", batchComputeEnvName)
 		batchComputeEnvArn = *batchComputeEnv.ComputeEnvironments[0].ComputeEnvironmentArn
 
+		// Launch template name
+		launchTemplateName, err := ensureLaunchTemplate(ec2Svc, userName, hostname)
+		if err != nil {
+			return "", err
+		}
```
[the remark this answers is not in the document]
Comment (jawadqur) [reply hatchery/nextflow.go]: I think we can, yes!

File: hatchery/nextflow.go [hunk 4]
```diff
@@ -526,13 +526,83 @@ func setupVpcAndSquid(ec2Svc *ec2.EC2, userName string, hostname string) (*strin
...
+			Config.Logger.Printf("Debug: Launch template '%s' does not exist, creating it", launchTemplateName)
+			launchTemplate, err := ec2Svc.CreateLaunchTemplate(&ec2.CreateLaunchTemplateInput{
+				LaunchTemplateName: aws.String(launchTemplateName),
+				LaunchTemplateData: &ec2.RequestLaunchTemplateData{
+					UserData: aws.String(userData),
+				},
+			})
+			if err != nil {
+				Config.Logger.Panicf("Error creating launch template '%s': %v", launchTemplateName, err)
+				return nil, err
```
[the remark this answers is not in the document]
Comment (jawadqur) [reply hatchery/nextflow.go]: Good catch, this was during debugging and I needed to exit and see the stacktrace.

File: hatchery/nextflow.go [hunk 5]
```diff
@@ -526,13 +526,83 @@ func setupVpcAndSquid(ec2Svc *ec2.EC2, userName string, hostname string) (*strin
...
+			}
+			Config.Logger.Printf("Debug: Created launch template '%s'", launchTemplateName)
+			return launchTemplate.LaunchTemplate.LaunchTemplateName, nil
+		} else {
+			Config.Logger.Printf("Error describing launch template '%s': %v", launchTemplateName, err)
+		}
+		return nil, err
+	}
+
+	if len(launchTemplate.LaunchTemplates) == 0 {
```
[the remark this answers is not in the document]
Comment (jawadqur) [reply hatchery/nextflow.go]: I thought so too initially, but it doesn't return an empty list, that function returns an "NotFoundException" instead. 

I think I have to refactor this code a bit to make it nicer, but handling the error code in the exception worked, but the length check of the returned list did not so maybe we can refactor this code a bit.

File: hatchery/nextflow.go (hunk 2 above)
Comment (jawadqur) [reply hatchery/nextflow.go:1552]: It's the marker for the start and end for the multiline string. Kind of similar to `cat << EOF`

**Review by paulineribeyre — APPROVED** [review]

**Pull request merged**